File: Android/Puzzlr/app/src/main/java/BlockChain/DataBlockchain.java
package BlockChain;

/**
 * Created by aniss on 09/03/16.
 */
import java.util.LinkedHashMap;
public class DataBlockchain implements IDataBlockchain {
    /** The blockchain. */
    OpenBlockchain blockchain;

    /** The type. */
    String type;

    /** The name. */
    String name;

    Hash hash;

    /**
     * Instantiates a new user blockchain.
     *
     * @param address
     *            the address
     * @param port
     *            the port
     */
    public DataBlockchain(String address, int port) {
        blockchain = new OpenBlockchain(address, port);

        type = "GOLANG";
        name = "de43e35f88de8c990310e91aefe3b91c01afffc548bca072070f2650165aef805f3ef7674696c8d84e4d572d420d0eb418477414dd76bac7748ef95cb5dba0fe";

        hash = new Hash();
    }

    /*
     * (non-Javadoc)
     *
     * @see imageblockchain.IImageBlockchain#sendMessage(java.lang.String,
     * java.lang.String)
     */
    @Override
    public void sendMessage(String username, String data) {

        String hashValue = hash.getSHA256(data);

        String[] argsUserMessages = { username, hashValue };
        String[] argsMessage = { hashValue, data };

        blockchain.invoke(type, name, "invoke", argsUserMessages);
        blockchain.invoke(type, name, "invoke", argsMessage);

    }

    /*
     * (non-Javadoc)
     *
     * @see imageblockchain.IImageBlockchain#getAllMessages(java.lang.String)
     */
    @Override
    public LinkedHashMap<String, String> getAllMessages(String username) {
        LinkedHashMap<String, String> listMessages = new LinkedHashMap<>();

        String rawHashes;

        String[] args = { username };

        JSONObject hashesJSON;

        hashesJSON = blockchain.query(type, name, "query", args);

        try {
            rawHashes = hashesJSON.getString("OK");

            // Loop on the hashes contained in an username.
            for (int i = 0; i < rawHashes.length() / 64; i++) {
                String hash = rawHashes.substring(i * 64, 64 * (i + 1));
                String[] arghash = { hash };
                JSONObject dataJSON = blockchain.query(type, name, "query", arghash);
                String value = dataJSON.getString("OK");
                listMessages.put(hash, value);

                //deleteData(hash);
            }
            //deleteData(username);

        } catch (JSONException e) {

            e.printStackTrace();
        }

        return listMessages;
    }

    private void deleteData(String key) {

        String[] args = { key };

        blockchain.invoke(type, name, "delete", args);

    }

}

File: iOS/Puzzlr/Hyperledger/DataBlockchain.swift
//
//  DataBlockchain.swift
//  Puzzlr
//
//  The blockchain which contains the data.
//
//  Created by Quentin Le Sceller on 09/03/2016.
//  Copyright © 2016 Quentin Le Sceller. All rights reserved.
//

import Foundation

class DataBlockchain {
    
    // The main blokchain:
    let blockchain : OpenBlockchain
    
    // The language used for the chaincode.
    let type : String
    
    // The name of the chaincode (usually hash value).
    let name : String
    
    // The hash value.
    let hash : Hash
    
    // Initializer
    init(ip: String,port : Int) {
        blockchain = OpenBlockchain(ip: ip, port: port)
        type = "GOLANG";
        name = "1ce07c739447a8abea238dfe9dc90b94bc32d56c2ea2f81af773cc11e9bfc9fc4b68a7cd7fccd418cd5da90fdfa6a72e5f72357bfc7b96ac8af251fd5f746a42";
        
        hash = Hash()
    }
    
    /**
     Send Message to the blockchain
     
     - parameter username: the username of the recipient
     - parameter data:     the data to send
     - parameter sync:     synchronous or asynchronous request
     */
    func sendMessage(username : String, data : String, sync : Bool){
        let hashValue = hash.getSHA256(data)
        print(hashValue)
        let argsUserMessages = [username, String(hashValue)]
        let argsMessage = [String(hashValue), data]
        
        blockchain.invoke(type, name: name, function: "invoke", args: argsUserMessages, sync: sync)
        blockchain.invoke(type, name: name, function: "invoke", args: argsMessage, sync: sync)
    }
    
    /**
     Get all the messages sent to an user.
     
     - parameter username: the username of the user.
     - parameter sync:     synchronous or asynchronous request
     
     - returns: all the messages in a dictionnary.
     */
    func getAllMessages(username : String, sync : Bool) -> [String : String] {
        var messages = [String : String]()
        
        let args = [username]
        
        let hashesJSON = blockchain.query(type, name: name, function: "query", args: args, sync: sync)
        
        var rawHashes = hashesJSON.substringFromIndex(7)
        let indexHash = rawHashes.endIndex.advancedBy(-2)
        rawHashes = rawHashes.substringToIndex(indexHash)
        
        for i in 0 ..< rawHashes.characters.count/64 {
            let hash = rawHashes.substringWithRange(Range<String.Index>(start: rawHashes.startIndex.advancedBy(64*i), end: rawHashes.startIndex.advancedBy(64*(i+1))))
            let argHash = [hash]
            
            let dataJSON = blockchain.query(type, name: name, function: "query", args: argHash, sync: sync)
            
            var value = dataJSON.substringFromIndex(7)
            let indexValue = value.endIndex.advancedBy(-2)
            value = value.substringToIndex(indexValue)
            
            messages[hash] = value
            deleteData(hash,sync: sync)
        }
        deleteData(username,sync: sync)
        
        return messages
    }
    
    /**
     Delete data from server
     
     - parameter key:  the key
     - parameter sync:     synchronous or asynchronous request
     */
    func deleteData(key : String, sync :Bool) {
        let args = [key]
        blockchain.invoke(type, name: name, function: "delete", args: args, sync: sync)
    }
    
}